### ISSUE TITLE
Method cascading (like Dart) -- enables Fluent interfaces without any extra effort

### ISSUE THREAD

**Issue by bandleader** [dotnet/vblang#245]
`object.Method()` -- returns result of `Method()`
`object..Method()` -- discards result of `Method()` (if any) and returns `object`
Therefore, you can use method chaining: `object..Method()..SecondMethod()..ThirdMethod()` ...without having to create specific Fluent-style methods that return the object.

# More information at the Dart website
- https://news.dartlang.org/2012/02/method-cascades-in-dart-posted-by-gilad.html
- Also: https://www.dartlang.org/guides/language/language-tour#cascade-notation-

# Benefits
- Benefit for class developers -- no need to specially create specific Fluent-style methods; simpler code.
- Benefit for class consumers -- ability to use Fluent-style access on classes that have not been coded for it.

# Properties
- Dart also allows setting properties with the same syntax, e.g. `object..Property = "value"` and then on the next line `..NextProperty = "value"`. 
- This may not fit in with VB being statement-separated. 
- We could do it the same way and decide that `..` at the beginning of a line continues the previous line. _(In any case these days VB's allow unfinished statements to implicitly continue onto new lines. This would be the same thing the other way around.)_
- Or, we could use a different syntax: `object..Property("value")..Property("value")`. It isn't as clear to the newcomer that this is a property assignment, but it does preserves the one-statement-per-line rule. _(Parameters for parameterized properties would be in the parentheses before the new value.)_

Thanks for such a great powerful and yet friendly language. Looking forward to the #discussion...

**Comment by ericmutta** [comment]
> @bandleader Benefit for class consumers -- ability to use Fluent-style access on classes that have not been coded for it.

Hmmm...is that really possible? In my experience with fluent-style classes, they are **specifically designed** to "accumulate" state then do something only at the end (e.g when you call something like a `build` function). For example this [AlertDialog.Builder](https://developer.android.com/reference/android/app/AlertDialog.Builder.html) in Android: you would chain a bunch of setXXX() methods and at the end call [create()](https://developer.android.com/reference/android/app/AlertDialog.Builder.html#create()).

The only classes I can think of that are NOT specifically designed for fluent-access but can still  be used in a fluent way are collection classes (e.g. using the syntax you suggested you would do `list..Add(1)..Add(2)..Add(3)`...which would work but is not as easy as using [collection initialisers](https://docs.microsoft.com/en-us/dotnet/visual-basic/programming-guide/language-features/collection-initializers/): `New List(Of Integer) From {1,2,3}` or simply `list.AddRange({1,2,3}))`

In any case, this would require more flexibility in placing newlines before/after operators (something we were talking about in #232), to avoid long method chains scrolling off to the right edge of the screen 👍

**Comment by reduckted** [comment]
>  In my experience with fluent-style classes, they are specifically designed to "accumulate" state then do something only at the end

Yep. Just look at any of the dependency injection container registration methods for an example. Using Autofac as an example, the `ContainerBuilder` class has a `Register` method. That returns a different type of object that lets you specify how to register the type, and calling methods on that object returns another type that lets you further configure the registration. It wouldn't make any sense to have all of those methods available from the `ContainerBuilder`.

@bandleader, do you have a real-life example of where this feature would be useful? That might help us understand the benefits of this.

**Comment by bandleader** [comment]
@ericmutta Depends. 90% of the time, the Fluent methods directly mutate state inside the object, instead of returning a new object. (I'm sure this is the case in the Android dialog builder that you linked as well.) Then they simply `Return Me` (or `return this;`) so that you can continue accessing the object (calling `.Create()`, or setting more properties. In these cases, the entire existence of all the setter methods is unnecessary, when simply setting properties directly would be fine, if we had a `..` operator to enable doing it in one line.

In 10% of cases (i.e. LINQ), the Fluent methods don't mutate their object, but instead return a new object. This is so you can compose them outward and retain the original in a different variable so you can compose it in a different direction. However, in these cases, the class consumer doesn't need/want this method-cascading operator, as you _want_ the reference returned by the method. As well, the class writer doesn't need our operator, because in any case he is defining some special functionality and returning a new object -- an obvious choice for a method. This is no different then `myString.Trim().Substring(1,3)`.

Rather, this operator is concerned with the 90% of cases (in my experience) when the entire Flulent method is only boilerplate for setting a property and returning the object so you can continue: `Function SetName(name As String) As MyObject: Me.name = name: Return Me: End Function` which could be eliminated if our operator provided access to the `name` property without breaking the chain.

Again, this isn't a novel idea, this is already there in Dart and Smalltalk. 

Happy to answer more questions as needed.

**Comment by reduckted** [comment]
> 90% of the time, the Fluent methods directly mutate state inside the object, instead of returning a new object.

Yes, that's often how it's implemented, but the important point is that the method that just returns `Me` has a return type of a certain interface that is _usually_ a different type. For example:

```vb
Interface IBuilder
    Function Register(t As Type) As IRegisteredTypeBuilder
Function 
```
> `Function SetName(name As String) As MyObject: Me.name = name: Return Me: End Function`

Can you provide an example of how you would set a property using this new syntax that you're proposing?

**Comment by bandleader** [comment]
> @reduckted  but the important point is that the method that just returns Me has a return type of a certain interface that is usually a different type. 
Point taken -- in that case they will simply use method chaining and not use our feature (method cascading).
I do believe that method cascading still has great value for the _x_ % of cases in which we return the object as the same type, to enable the next method call.

> @reduckted Can you provide an example of how you would set a property using this new syntax that you're proposing? [...] Ah, is the first point under "Properties" in your opening statement how you are proposing to do it? I suspect that would end up being ambiguous if we allow implicit line continuations before a period.

I much prefer (for that and other reasons) the `obj..Prop("value")` syntax. Much cleaner as well, and more in tune with BASIC's one-statement-per-line.
I am not bothered by the fact that `Prop` above can be either a property or method.
_(It's no worse than `obj.fld` which can be either a field, property, or argumentless method, and also `obj.method()` which can be either a Function or Sub.)_

**Comment by xieguigang** [comment]
@bandleader, @reduckted 

Fluent-style method cascading like [**d3js.labeler**](https://github.com/xieguigang/sciBASIC/blob/master/Data_science/Mathematica/Plot/Plots-statistics/RegressionPlot.vb#L156)

```vbnet
Call d3js.labeler _
    .Labels(labels) _
    .Anchors(labels.GetLabelAnchors(pointSize)) _
    .Width(rect.Width) _
    .Height(rect.Height) _
    .Start(showProgress:=False)
```

can implements by using ``With`` keyword in VB.NET:

```vbnet
With d3js.labeler 
    .Labels(labels) 
    .Anchors(labels.GetLabelAnchors(pointSize)) 
    .Width(rect.Width) 
    .Height(rect.Height) 
    .Start(showProgress:=False)
End With
```

> Yes, that's often how it's implemented, but the important point is that the method that just returns Me has a return type of a certain interface that is usually a different type.

Using a more simple version ``With`` syntax for implements a Fluent-style method cascading **even though the target method is a ``Sub``(can not returns value)**, without change any code that you wrote before:

```vbnet
d3js.labeler With
    .Labels(labels) 
    .Anchors(labels.GetLabelAnchors(pointSize)) 
    .Width(rect.Width) 
    .Height(rect.Height) 
    .Start(showProgress:=False)
' An empty line or single line code comment will break this method cascading
```

**Comment by bandleader** [comment]
@xieguigang Thank you for your comment. The `With` construct, while well-known and useful, IMHO does not replace a true method-cascading operator:

- Cannot be on a single line
- Moreover, requires a block, including a start line `With obj` and an end line `End With`
- This often negates whatever readability/brevity benefits it could provided
- It also does not return an expression; you would have to instantiate the `d3js.labeler` separately in advance and assign it to a variable. Whereas with method cascading, your entire code (instantiation, settings and all) could be written simply as: `Dim labeler = New d3js.Labeler()..Labels(labels)..OtherMethods(whatever)..Start(showProgress:=False)`

As far as the "simple version" you suggested: it simply removes the `End With` but requires a blank line, so it's not much better -- but either way it is not BASIC-like syntax (more Python-like), and @AnthonyDGreen and the rest of the VB team will not accept that.

**Comment by xieguigang** [comment]
Place all of the method cascading in one single line is not recommended, as it:

+ not suitable for print your code on an A4 paper for discussion with your colleague 
+ inconvenient for your colleague reading your code on the screen when your code is too long, you needs a larger screen for display the entire line of your code

**Comment by bandleader** [comment]
@xieguigang Understood, but that's not a reason to *force* developers to split it into multiple lines, when they may have many good reasons to keep it on one line (I certainly do).

**Comment by bandleader** [comment]
As some have requested a real-life example:

I have a Fluently-designed class called `MailerMessage` -- an advanced helper cless to construct email messages in both HTML and text formats, send them using a separate `Mailer` object (which uses multiple SMTP servers, etc.), and log them.
```vb
    Class MailerMessage
            Inherits System.Net.Mail.MailMessage
            Private DebugDetails As String = ""
            Private BodyHTML As String = ""
            Private BodyPlainText As String = ""
            Public Mailer As Mailer

            Public Function BodyAppend(...) As MailerMessage
                'Code to intelligently build the message body...
                Return Me
            End Function

            'This method does nothing but set a property! 
            '...but is necessary to continue the Fluent method chain.
            'There are MANY like it...
            Public Function Subject(subj As String) As MailerMessage  
                Me.Subject = subj
                Return Me
            End Function

            Public Function Send() As MailerMessage
                Me.Mailer.Send(Me)
                Return Me
            End Sub

            Public Function LogSuccess() As MailerMessage
                'Code to log our successfully sent message
                Return Me
            End Sub

            '...and MANY, MANY more methods (omitted for brevity) that help you 
            'easily add recipients, attachments, HTML, plain text, and much more. 
            'Each one returns 'Me' so that you can instantiate a simple message, 
            'send and log it - all in a single statement.
   End Class

   '=== EXAMPLE USAGE: (user.NewMessage returns a pre-addressed MailerMessage)
   user.NewMessage().Subject("Thank you for joining GitHub!").BodyAppend("message goes here").Send().LogSuccess()
```

### Great Fluent class. However:

1. I had to explicitly design my class to be Fluent. I converted every Sub into a Function that returns `MailerMessage`, and added `Return Me` to the end, as well as every time I wanted to `Exit Sub`, etc.
2. I had to create a method for every single property! See the `Subject` method above. With method cascading, I could simply use `..Subject("Subject here")` and method cascading would use the already-implemented property.
3. The class may have been written by someone else, with no Fluent API, and over which I have no control. With method cascading, I could still use his API in a Fluent manner.
4. In fact, I could perhaps have used `Net.MailMessage` and not written this class at all! (At most, a couple of extension methods to `Net.MailMessage` could have done the trick!)

I hope this demonstrates the utility of the feature! Would love input from @KathleenDollard or any members of the VB team.

**Comment by AnthonyDGreen** [comment]
I love the idea, the big concern I have is that VB already has the `.` (simple) and `...` (XML) member-access operators. Which are far enough apart right now, but adding `..` might suggest some kind of linear relationship between the three (maybe). But that might not matter in the end.
